guides/contexts.md

```diff
@@ -8,1060 +8,1313 @@
+
+With our schema associations set up, we can implement the selection of categories in our product form. To do so, we need to translate the user input of catalog IDs from the front-end to our many-to-many association. Fortunately Ecto makes this a breeze now that our schema is set up. Open up your catalog context and make the following changes:
+
 
 ```elixir
-+ alias Hello.Accounts.User
+- def get_product!(id), do: Repo.get!(Product, id)
++ def get_product!(id) do
++   Product |> Repo.get(id) |> Repo.preload(:categories)
++ end
+
+  def create_product(attrs \\ %{}) do
+    %Product{}
+-   |> Product.changeset(attrs)
++   |> change_product(attrs)
+    |> Repo.insert()
+  end
 
+  def update_product(%Product{} = product, attrs) do
+    product
+-   |> Product.changeset(attrs)
++   |> change_product(attrs)
+    |> Repo.update()
+  end
 
-  schema "credentials" do
-    field :email, :string
--   field :user_id, :id
-+   belongs_to :user, User
+  def change_product(%Product{} = product, attrs \\ %{}) do
+-   Product.changeset(product, attrs)
++   categories = list_categories_by_id(attrs["category_ids"])
 
-    timestamps()
++   product
++   |> Repo.preload(:categories)
++   |> Product.changeset(attrs)
++   |> Ecto.Changeset.put_assoc(:categories, categories)
   end
 
++ def list_categories_by_id(nil), do: []
++ def list_categories_by_id(category_ids) do
++   Repo.all(from c in Category, where: c.id in ^category_ids)
++ end
 ```
 
-We used the `belongs_to` macro to map our child relationship to the parent `User`. With our schema associations set up, let's open up `accounts.ex` and make the following changes to the generated `list_users` and `get_user!`  functions:
+First, we added `Repo.preload` to preload our categories when we fetch a product. This will allow us to reference `product.categories` in our controllers, templates, and anywhere else we want to make use of category information. Next, we modified our `create_product` and `update_product` functions to call into our existing `change_product` function to produce a changeset. Within `change_product` we added a lookup to find all categories if the `"category_ids"` attribute is present. Then we preloaded categories and called `Ecto.Changeset.put_assoc` to place the fetched categories into the changeset. Finally, we implemented the `list_categories_by_id/1` function to query the categories matching the category IDs, or return an empty list if no `"category_ids"` attribute is present. Now our `create_product` and `update_product` functions receive a changeset with the category associations all ready to go once we attempt an insert or update against our repo.
+
+Next, let's expose our new feature to the web by adding the category input to our product form. To keep our form template tidy, let's write a new function to wrap up the details of rendering a category select input for our product. Open up your `ProductView` in lib/hello_web/views/product_view.ex`: and key this in:
```

Comment:
A "`" is missing before "lib/hello_web/views/product_view.ex". It results in a bad formatted doc.